### PR TITLE
ROX-20905: Handle multiple yaml files

### PR DESCRIPTION
**Pull request by jschnath**
## Description

Take multiple yaml files as input to roxctl deployment check --file instead of only a single one.
In order not to break legacy support I decided to add an additional --files flag with this behavior. Let me know in the comments if it would make more sense to also remove the old --file flag, OR to make the --file flag work in the way the --files flag works now.

On a sidenote I also moved the assignment of random IDs for deployments parsed from yaml from the previous PR https://github.com/stackrox/stackrox/pull/8763 to a different spot where I think it makes more sense.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

I manually tested roxctl deployment check --files with multiple input files and some print statements and it looks to be behaving as expected to me.

### CODE REVIEW

**Comment by dhaus67** [comment]
Is this reviewable already @jschnath ? If not, let's convert this to a draft to not ping code owners _yet_.

**Comment by jschnath** [comment]
@dhaus67 I needed it as a full PR so the image would be pushed to Prow. But at this point it is also reviewable. I just got done manually testing the change on a cluster and will now write in PR post.

**Comment by jschnath** [comment]
On a sidenote, maple is now working on a feature related to roxctl https://issues.redhat.com/browse/ROX-11350, and we will be getting to you soon with some design ideas, and we might need some help with PR reviews.

File: roxctl/deployment/check/check.go
Comment (fredrb) [inline roxctl/deployment/check/check.go:257]: Isn't it possible to have two entries in `alerts` that have the same entity (i.e. same deployment)? I believe this would cause every deployment entity to get assigned a new ID regardless of being the same entity. 

This could be related to the failing unit tests (did not check this): https://github.com/stackrox/stackrox/actions/runs/7035940956/job/19147480721?pr=8822

File: roxctl/deployment/check/check.go
Comment (fredrb) [inline roxctl/deployment/check/check.go:255]:
```suggestion
	// get a Kubernetes generated ID. This is a temporary ID only required for roxctl to distinguish
```

File: roxctl/deployment/check/check.go
Comment (fredrb) [inline roxctl/deployment/check/check.go:174]: You probably want to use `errorhelpers.ErrorList` to check all files before returning the error

File: roxctl/deployment/check/check.go
Comment (fredrb) [inline roxctl/deployment/check/check.go:112]: I don't understand why keeping both the singular `--file` and adding a new `--files` flag. Shouldn't single files be handled as a single element in `--files` parameter?

File: roxctl/deployment/check/check.go
Comment (fredrb) [inline roxctl/deployment/check/check.go:168]: Another reason to remove `--file` in favor of `--files` only is to avoid this odd behavior if both are passed but only one considered.

File: roxctl/deployment/check/check.go
Comment (fredrb) [inline roxctl/deployment/check/check.go:215]: You probably want to add a line break in the beginning as well in case the input has no endline. 
```suggestion
				deploymentFileContents = append(deploymentFileContents, "\n---\n"...)
```

**Review by fredrb — CHANGES_REQUESTED** [review]
I'm missing some unit test changes in this PR. Could you please check the tests that already exist for roxctl and update them according to the new logic. 

Also, consider the idea of removing `--file` in favor or `--files`. 

One downside I just realized for appending the files together as opposed to sending multiple files is that if one file is formatted incorrectly you jeopardize the entire check by making the entire yaml invalid.

File: roxctl/deployment/check/check.go
Comment (jschnath) [reply roxctl/deployment/check/check.go:112]: See the post in the PR, I did not want to break existing scripts of customers using the `--file` flag if possible. If we combine the two perhaps it would make more sense under `--file`, but that also makes it not obvious that it can accept multiple files. Thoughts?

File: roxctl/deployment/check/check.go
Comment (jschnath) [reply roxctl/deployment/check/check.go:215]: Is a `.yaml` still formatted correctly if there's an empty line before the `---` separating two deployments? I might need to test this, and if that's the case perhaps check whether input files have a newline at EOF.

**Comment by jschnath** [comment]
@fredrb I'll add some test coverage and then mark you for re-review.

File: roxctl/deployment/check/check.go
Comment (fredrb) [reply roxctl/deployment/check/check.go:215]: I think it should be fine. Adding multiple lines between separators [here](https://www.yamllint.com/) seems to result in a valid yaml.

File: roxctl/deployment/check/check.go
Comment (fredrb) [reply roxctl/deployment/check/check.go:112]: What is the behavior of the slice? If the flag usage for a single element is identical to string argument then they are compatible and you can simply swap the flag. In this case, drop the `--files` and let `--file` accept multiple files.

File: roxctl/deployment/check/check.go
Comment (fredrb) [reply roxctl/deployment/check/check.go:168]: See comment above, I think there is still a reason to merge this. Otherwise, we should at least show a warning if both parameters are used together.

File: roxctl/deployment/check/check.go
Comment (jschnath) [reply roxctl/deployment/check/check.go:257]: I think you're right after some testing. I'll apply the same random ID to multiple deployments within the same alert then.

File: roxctl/deployment/check/check.go
Comment (Maddosaurus) [reply roxctl/deployment/check/check.go:112]: I know that `roxctl` is used often to automate tasks around ACS, so I'd be strongly against **removing** a parameter to not break existing automation of customers.

How about we take inspiration from `kubectl` and `helm`?
`Helm` is using the notation @janisz described: 
	`helm install ./path --values ./internalValues.yaml --values ./customSettings.yaml`
Whereas `kubectl` can apply a whole folder by providing `kubectl apply -f` with a directory instead of a file. It also supports `--recursive` to also apply all files in any subfolders of the given path.

Either way we wouldn't break existing functionality, and still be able to supply multiple files.
I'd be in favor of the Helm way. 

Here is some discussion involving how to do this in [Cobra](https://github.com/spf13/cobra/issues/661) and with the [flags package](https://stackoverflow.com/questions/28322997/how-to-get-a-list-of-values-into-a-flag-in-golang/28323276#28323276) that I dug up out of interest.

**Comment by jschnath** [comment]
I just got around to addressing all the feedback. It took longer than expected since I was sick last week.

File: roxctl/deployment/check/check.go
Comment (jschnath) [reply roxctl/deployment/check/check.go]: Done

File: roxctl/deployment/check/check.go
Comment (jschnath) [reply roxctl/deployment/check/check.go]: Good to know, I used it now.

File: roxctl/deployment/check/check.go
Comment (jschnath) [reply roxctl/deployment/check/check.go]: I ended up merging the two of them into `--file`, preserving old behavior.

**Comment by jschnath** [comment]
`gke-qa-e2e-tests` failure looks unrelated to my changes and has been very flaky recently. `gke-sensor-integration-tests` shows some odd timeout in the logs, I think I'll just rerun it.

File: roxctl/deployment/check/check_test.go
Comment (fredrb) [inline roxctl/deployment/check/check_test.go:250]: Let's avoid suffixing variables with numbers. Maybe `mockDetectionService`?

File: roxctl/deployment/check/check.go
Comment (fredrb) [inline roxctl/deployment/check/check.go:131]: Why is the slice notation used here?
```suggestion
	deploymentCheckCmd.joinedFiles = strings.Join(deploymentCheckCmd.files, ",")
```

File: roxctl/deployment/check/check.go
Comment (jschnath) [reply roxctl/deployment/check/check.go:216]: To collect metrics that it's not growing too large, good idea. I'll add it.

File: roxctl/deployment/check/check_test.go
Comment (jschnath) [reply roxctl/deployment/check/check_test.go:287]: You're right, I changed it and the tests still pass.

File: roxctl/deployment/check/check_test.go
Comment (jschnath) [reply roxctl/deployment/check/check_test.go:302]: Ah, so that's the difference. I think there's a few other instances where I asserted as well, I'll look for those too.

File: roxctl/deployment/check/check_test.go
Comment (jschnath) [reply roxctl/deployment/check/check_test.go:308]: Gotcha

File: roxctl/deployment/check/check_test.go
Comment (jschnath) [reply roxctl/deployment/check/check_test.go:309]: It's formatted this way somewhere within `server.request.String()`, but I gave up tracking down where exactly after going through 10 layers of proto functions and branches. Note that the `strconv.Quote()` call is important too for the string to be formatted the same way as the actual output.

Presumably whoever consumes the output of `server.request.String()` expects it in this format.

File: roxctl/deployment/check/check.go
Comment (jschnath) [reply roxctl/deployment/check/check.go:131]: I thought it was necessary, but now that I tested it leaving it out the slice notation leads to the same result. I've removed it now.

File: roxctl/deployment/check/check_test.go
Comment (jschnath) [reply roxctl/deployment/check/check_test.go:250]: Okay, done.

File: roxctl/deployment/check/check.go
Comment (Maddosaurus) [reply roxctl/deployment/check/check.go:215]: Would it make a difference to always append this to the YAML?
I usually start all my YAML files with `---`, just to be on the safe side. Maybe we can simplify here, or is there a problem with a leading newline as first line in a file?

File: roxctl/deployment/check/check.go
Comment (dhaus67) [inline roxctl/deployment/check/check.go:213]: Haven't seen a test case for it, so might be worthwhile checking whether two separators in a row are seen as valid, i.e.
```yaml
something: test
---
---
another-thing: test
```

File: roxctl/deployment/check/check.go
Comment (dhaus67) [inline roxctl/deployment/check/check.go:217]: A couple of things with this:
- you are using `Info` instead of `Infof`.
- you are using `%s` for a number, which should be `%d`.
- I don't really think the file size is a meaningful information to provide to the user. The file names could make sense, but other than that I'd rather drop it.

File: roxctl/deployment/check/check_test.go
Comment (dhaus67) [reply roxctl/deployment/check/check_test.go:303]: It's the proto field from the request, what you ought to be doing is comparing the string with `server.request.GetYaml()`

**Review by dhaus67 — CHANGES_REQUESTED** [review]

File: roxctl/deployment/check/check.go
Comment (jschnath) [reply roxctl/deployment/check/check.go:140]: Fair points, and I'm a bit out of my depth when it comes to our Sarif printers, so great if Merlin can handle that, however.
> use the first image in the list of files instead. It'd potentially be a "wrong" reference, but at the very least not a invalid reference.
This is actually the first workaround I tried but it does not work since `deploymentCheckCmd.files` is empty at the time it is passed to `NewSarifPrinterFactory`, so pointing at the first element of the slice is out of range. I could create a `deploymentCheckCmd.firstFile` of type string and point at that, then populate it after flags are set as a workaround, what do you think?

File: roxctl/deployment/check/check.go
Comment (jschnath) [reply roxctl/deployment/check/check.go:133]: Ah, I did not test with `-o sarif`. It seems that I'll be removing the `joinedFiles` field, but I'll still need to find the right spot to set a different parameter such as `firstFile` after flag values are populated from the user, if I go with the approach I suggest in the comment below.

File: roxctl/deployment/check/check.go
Comment (jschnath) [reply roxctl/deployment/check/check.go:175]: Fair, done.

File: roxctl/deployment/check/check.go
Comment (jschnath) [reply roxctl/deployment/check/check.go:213]: The only way that would happen is if the user added an empty yml file, is that a valid use-case or should we throw an Error? eg: `roxctl deployment check --file deploy1.yml --file emptydeploy.yml --file deploy2.yml` resulting in
`
deploy1: test
---
---
deploy2: test
`

File: roxctl/deployment/check/check.go
Comment (jschnath) [reply roxctl/deployment/check/check.go:217]: Oops, I'll fix the `%d` and the `Infof`. In terms of usage, the idea behind logging this is a concern that if we append many deployments into one the resulting concatenation could become too large. By logging this we could stay on top of this. At least I think that's what the idea behind this was, thoughts @fredrb?

File: roxctl/deployment/check/check.go
Comment (jschnath) [reply roxctl/deployment/check/check.go:41]: > The environment has a logger attached to it that you should use: d.env.Logger().InfofLn

Oh even better, I'll use that one then :+1:

File: roxctl/deployment/check/check.go
Comment (jschnath) [reply roxctl/deployment/check/check.go:215]: I think I need to spin up a cluster and do some testing regarding leading newlines in yaml and multiple lines of `---` as mentioned in a comment below

File: roxctl/deployment/check/check_test.go
Comment (jschnath) [reply roxctl/deployment/check/check_test.go]: Nice one, done

File: roxctl/deployment/check/check.go
Comment (jschnath) [reply roxctl/deployment/check/check.go:41]: Using `d.env.Logger().InfofLn` has the unintended side-effect of making most tests fail, since they assert against what messages are logged. I'll need to either add that line to the tests or remove the log after all.

File: roxctl/deployment/check/check.go
Comment (fredrb) [reply roxctl/deployment/check/check.go:217]: The reason why I asked for this is because we have seen issues elsewhere where things fail due to large payloads and we can't really pinpoint which message was it. Logging the size of this file can help us in the future if users start complaining that deployment check is failing. 

It's a fair point that this information isn't meaningful to users, unless they want to open a ticket. 

Do we have debug logs in roxctl? If yes, perhaps let's add it as debug level. If users run into failures here we can ask them to run the command with debug logs enabled so we can see how large is the file roxctl is trying to send.

File: roxctl/deployment/check/check.go
Comment (fredrb) [reply roxctl/deployment/check/check.go:41]: Changing the level of the log message you added to `DEBUG` likely would solve this as well.

File: roxctl/deployment/check/check_test.go
Comment (fredrb) [inline roxctl/deployment/check/check_test.go:300]: You don't need to define this here, just use the declaration + assignment notation below (`:=`):
```
expectedBinary, err := os.ReadFile("testdata/testMultipleFilesExpectedYaml.yaml")
```

File: roxctl/deployment/check/check_test.go
Comment (fredrb) [inline roxctl/deployment/check/check_test.go:369]: nit: s/file/files

File: roxctl/deployment/check/check.go
Comment (fredrb) [reply roxctl/deployment/check/check.go:213]: I tested this and it is indeed a problem. If a file ends with a separator (`---`) it fails if there is another file after it. It works if checked individually or as the last file in the sequence.

This needs to be fixed before merging.

**Review by fredrb — COMMENTED** [review]
Thanks for addressing all my original comments. There are two points from @dhaus67 review above that need to be fixed before this is merged:

1) The log line: my suggestion is to change it to debug level
2) Yaml files ending with separators. If two files work in the deployment check individually, they should also work when passed together.

File: roxctl/deployment/check/check.go
Comment (jschnath) [reply roxctl/deployment/check/check.go:41]: There is no debug log level in the logger we use there, and now we have agreed on just removing the log line.

File: roxctl/deployment/check/check.go
Comment (jschnath) [reply roxctl/deployment/check/check.go]: My latest push fixes this by normalizing parsed YAML files

File: roxctl/deployment/check/check_test.go
Comment (jschnath) [reply roxctl/deployment/check/check_test.go:330]: I thought this would not work because err was already declared above, but it did work.

File: roxctl/deployment/check/check_test.go
Comment (jschnath) [reply roxctl/deployment/check/check_test.go:399]: Oh, must've missed that one when refactoring. Changing it.

File: roxctl/deployment/check/check.go
Comment (jschnath) [reply roxctl/deployment/check/check.go:140]: I've gone ahead and created the issue for SARIF printers https://issues.redhat.com/browse/ROX-21443
TODO @stackrox/merlin

File: roxctl/deployment/check/check.go
Comment (dhaus67) [reply roxctl/deployment/check/check.go:41]: Let's remove this here then?

File: roxctl/deployment/check/check_test.go
Comment (dhaus67) [inline roxctl/deployment/check/check_test.go:284]: Just as a comment, feel free to ignore: I usually prefer using backsticks compared to double quotes for multiine strings, this makes things more readable to me.

File: roxctl/deployment/check/check_test.go
Comment (dhaus67) [inline roxctl/deployment/check/check_test.go:321]: super-nit:
```suggestion
	conn, closeF, server := d.createGRPCMockDetectionService(testDeploymentAlertsWithoutFailure, nil)
```

**Review by dhaus67 — COMMENTED** [review]
Minor comments, otherwise LGTM.

**Review by dhaus67 — APPROVED** [review]

File: roxctl/deployment/check/check.go
Comment (jschnath) [reply roxctl/deployment/check/check.go:89]: Done

File: roxctl/deployment/check/check_test.go
Comment (jschnath) [reply roxctl/deployment/check/check_test.go:284]: I prefer it as is

File: roxctl/deployment/check/check_test.go
Comment (jschnath) [reply roxctl/deployment/check/check_test.go:321]: Done

**Review by fredrb — APPROVED** [review]
Thanks for addressing all the feedback. Could you try to rebase to see if the `Lint` step passes?

**Comment by jschnath** [comment]
> Could you try to rebase to see if the Lint step passes?

Sure

**Pull request merged**
